src/rust/engine/process_execution/src/local.rs

```diff
@@ -114,6 +115,79 @@ impl CommandRunner {
   }
 }
 
+lazy_static! {
+  static ref IS_LIKELY_IN_DOCKER: bool = is_likely_in_docker().unwrap_or_else(|e| {
+    log::info!("Failed to detect whether we are running in docker: {}", e);
+    false
+  });
+}
+
+///
+/// Attempts to detect whether we are running inside a docker container.
+///
+/// NB: Do not call this directly: it is stored in the IS_LIKELY_IN_DOCKER lazy_static.
+///
+fn is_likely_in_docker() -> Result<bool, String> {
+  if cfg!(not(target_os = "linux")) {
+    return Ok(false);
+  }
+
+  // Attempt to detect whether we are in docker. See:
+  //   https://stackoverflow.com/questions/20010199/how-to-determine-if-a-process-runs-inside-lxc-docker
+  let cgroups_matches = {
+    BufReader::new(
+      std::fs::File::open("/proc/1/cgroup")
+        .map_err(|e| format!("Failed to inspect `/proc` to detect docker: {}", e))?,
+    )
+    .lines()
+    .filter_map(|line_result| match line_result {
+      Ok(line) if line.contains("docker") || line.contains("lxc") => Some(line),
+      _ => None,
+    })
+    .collect::<Vec<_>>()
+  };
+
+  if cgroups_matches.is_empty() {
+    Ok(false)
+  } else {
+    log::debug!(
+      "Detected potential docker container based on cgroups ({}): will `sync` \
+      before executing processes.",
+      cgroups_matches.join(", ")
+    );
+    Ok(true)
+  }
+}
+
+///
+/// If we are potentially running inside a docker container (TODO: technically only `aufs` is
+/// relevant), `sync` the filesystem. Noop on other platforms.
+///
+/// See https://github.com/moby/moby/issues/9547.
+///
+async fn sync_if_needed() -> Result<(), String> {
+  if !(*IS_LIKELY_IN_DOCKER) {
+    return Ok(());
+  }
+
+  let output = Command::new("/bin/sync")
+    .stdin(Stdio::null())
+    .output()
+    .await
+    .map_err(|e| format!("Failed to spawn `sync` in likely docker container: {}", e))?;
+
+  if !output.status.success() {
+    let stdout = String::from_utf8_lossy(&output.stdout);
+    let stderr = String::from_utf8_lossy(&output.stderr);
+    return Err(format!(
+      "Failed to `sync` in likely docker container ({}): stdout: {}, stderr: {}",
```

Comment:
This might help with John's concern of the binary not being available.

```suggestion
      "Failed to run `/bin/sync` in likely docker container ({}): stdout: {}, stderr: {}",
```